gen/write.rs

```diff
@@ -300,6 +321,21 @@ fn write_struct_using(out: &mut OutFile, ident: &Ident) {
     writeln!(out, "using {} = {};", ident, ident);
 }
 
+fn write_struct_with_methods(out: &mut OutFile, ety: &ExternType, methods: &Vec<&ExternFn>) {
+    for line in ety.doc.to_string().lines() {
+        writeln!(out, "//{}", line);
+    }
+    writeln!(out, "struct {} final {{", ety.ident);
+    for method in methods {
+        write!(out, "  ");
+        let sig = &method.sig;
+        let local_name = method.ident.to_string();
+        write_rust_function_shim_decl(out, &local_name, sig, None, false);
+        writeln!(out, ";");
+    }
+    writeln!(out, "}};");
```

Comment:
We need a private default constructor in here. Otherwise this allows C++ to obtain an opaque Rust type by value and call methods on it, which is unsound.

Specifically the R2 struct from tests/ffi/lib.rs currently turns into:

```cpp
struct R2 final {
  size_t get() noexcept;
  size_t set(size_t n) noexcept;
};
```

and this allows C++ to write `R2{}.get()` which becomes UB.

Reply:
Thanks for catching that!